### PR TITLE
chore: Add Hir -> Ast conversion

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/4796

## Summary\*

 Adds a way to go from the Hir back to the Ast.
This conversion isn't exact and information is lost currently. Notably many spans are lost and some constructs like `Type::TypeVariable` have no equivalent in the Ast.

This also isn't expected to be idempotent. Different nodes may be desugared, `let` statements may have explicit types added, etc.

## Additional Context

This is unused currently and can't be triggered from the current compiler at all. I'm adding this PR now to split off required changes for metaprogramming into smaller chunks.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/ast/mod.rs [hunk 1]
```diff
@@ -112,6 +112,9 @@ pub enum UnresolvedTypeData {
         /*env:*/ Box<UnresolvedType>,
     ),
 
+    // The type of quoted code for metaprogramming
+    Code,
```
Comment (jfecher) [inline compiler/noirc_frontend/src/ast/mod.rs:116]: We had a `Type::Code` but not a `UnresolvedTypeData::Code` before so I needed to have this to have something to convert `Type::Code` to.

File: compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+                let r#type = interner.id_type(let_stmt.expression).to_ast();
+                let expression = let_stmt.expression.to_ast(interner);
+                StatementKind::Let(LetStatement { pattern, r#type, expression })
+            }
+            HirStatement::Constrain(constrain) => {
+                let expr = constrain.0.to_ast(interner);
+                let message = constrain.2.map(|message| message.to_ast(interner));
+
+                // TODO: Find difference in usage between Assert & AssertEq
+                StatementKind::Constrain(ConstrainStatement(expr, message, ConstrainKind::Assert))
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs:42]: where does the second `ValueId` in the `Constrain` end up? i.e. `constrain.1`

File: compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            Type::Constant(_) => panic!("Type::Constant where a type was expected: {self:?}"),
+            Type::Code => UnresolvedTypeData::Code,
+            Type::Error => UnresolvedTypeData::Error,
+        };
+
+        UnresolvedType { typ, span: None }
+    }
+
+    fn to_type_expression(&self) -> UnresolvedTypeExpression {
+        let span = Span::default();
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs:290]: would it make sense to use `Option<Span>` in `UnresolvedTypeExpression` to distinguish these cases?

**Review by michaeljklein — CHANGES_REQUESTED** [review]

File: compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs (hunk 2 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs:42]: `constrain.1` is a FileId, not a ValueId

File: compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs (hunk 3 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/hir/comptime/hir_to_ast.rs:290]: The main problem is that we don't have spans for Types so we'd still have to choose either `None` or `Some(Span::default())` for everything. In the future we'll need to fix this I think but there isn't a satisfactory solution now.

**Comment by jfecher** [comment]
This PR should be ready for review now, CI is fixed

**Review by michaeljklein — APPROVED** [review]
LGTM

**Pull request merged**